extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaIndexTaskClient.java

```diff
@@ -143,147 +111,147 @@ public void resume(TaskLocation location, String id)
-          String.format("%s/%s/%s", BASE_PATH, id, "offsets/end"),
-          resume ? "resume=true" : null,
-          null
-      );
       submitRequest(
-          new Request(HttpMethod.POST, serviceUri.toURL()).setContent(
-              MediaType.APPLICATION_JSON,
-              jsonMapper.writeValueAsBytes(endOffsets)
-          ), true
+          id,
+          HttpMethod.POST,
+          "offsets/end",
+          resume ? "resume=true" : null,
+          jsonMapper.writeValueAsBytes(endOffsets),
+          true
       );
     }
-    catch (Exception e) {
+    catch (IOException e) {
       throw Throwables.propagate(e);
     }
   }
 
-  private StatusResponseHolder submitRequest(Request request, boolean retry)
+  @VisibleForTesting
+  RetryPolicyFactory createRetryPolicyFactory() {
+    return new RetryPolicyFactory(new RetryPolicyConfig()
+                                      .setMinWait(Period.seconds(2))
+                                      .setMaxWait(Period.seconds(8))
+                                      .setMaxRetryCount(5)
+    );
+  }
+
+  @VisibleForTesting
+  void checkConnection(String host, int port) throws IOException
+  {
+    new Socket(host, port).close();
+  }
+
+  private FullResponseHolder submitRequest(String id, HttpMethod method, String pathSuffix, String query, boolean retry)
+  {
+    return submitRequest(id, method, pathSuffix, query, new byte[0], retry);
+  }
+
+  private FullResponseHolder submitRequest(
+      String id,
+      HttpMethod method,
+      String pathSuffix,
+      String query,
+      byte[] content,
+      boolean retry
+  )
   {
     final RetryPolicy retryPolicy = retryPolicyFactory.makeRetryPolicy();
     while (true) {
-      StatusResponseHolder response = null;
+      FullResponseHolder response = null;
+      Request request = null;
 
       try {
+        TaskLocation location = taskLocationProvider.getTaskLocation(id);
+        if (location == null) {
+          throw new NoTaskLocationException();
+        }
+
         // Netty throws some annoying exceptions if a connection can't be opened, which happens relatively frequently
         // for tasks that happen to still be starting up, so test the connection first to keep the logs clean.
-        new Socket(request.getUrl().getHost(), request.getUrl().getPort()).close();
+        checkConnection(location.getHost(), location.getPort());
 
         try {
-          response = httpClient.go(request, new StatusResponseHandler(Charsets.UTF_8)).get();
+          URI serviceUri = new URI(
+              "http",
+              null,
+              location.getHost(),
+              location.getPort(),
+              String.format("%s/%s/%s", BASE_PATH, id, pathSuffix),
+              query,
+              null
+          );
+          request = new Request(method, serviceUri.toURL());
+          request.addHeader("X-Druid-Task-Id", id); // used to validate that we are talking to the correct worker
+
+          if (content.length > 0) {
```

Comment:
This is the request content and not the response content; if no content data was provided to the method, then it doesn't add it to the request.